tests/test_applets/batchProcessing/test_batchProcessingGui.py

```diff
@@ -0,0 +1,203 @@
+        success_args[0]()
+
+        assert_buttons_in_init_state(batch_processing_gui)
+
+        qtbot.mouseClick(batch_processing_gui.run_button, Qt.LeftButton, delay=1)
+        assert len(RequestMock._instances) == 2
+        assert_buttons_in_running_state(batch_processing_gui)
+
+        # cancel:
+        qtbot.mouseClick(batch_processing_gui.cancel_button, Qt.LeftButton, delay=1)
+        req = RequestMock._instances[1]
+        req.cancel.assert_called_once()
+        cancel_args, cancel_kwargs = req.notify_cancelled.call_args
+        assert len(cancel_args) == 1
+
+        info_messagebox = mock.Mock()
+        with mock.patch(
+            "ilastik.applets.batchProcessing.batchProcessingGui.QMessageBox.information", new=info_messagebox
+        ):
+            cancel_args[0]()
+            info_messagebox.assert_called_once()
+
+        assert_buttons_in_init_state(batch_processing_gui)
+
+        # failure:
+        qtbot.mouseClick(batch_processing_gui.run_button, Qt.LeftButton, delay=1)
+        assert len(RequestMock._instances) == 3
+        assert_buttons_in_running_state(batch_processing_gui)
+
+        req = RequestMock._instances[2]
+        failed_args, failed_kwargs = req.notify_failed.call_args
+        assert len(failed_args) == 1
+
+        critical_messagebox = mock.Mock()
+        with mock.patch(
+            "ilastik.applets.batchProcessing.batchProcessingGui.QMessageBox.critical", new=critical_messagebox
+        ):
+            failed_args[0](mock.MagicMock(Exception), None)
+            critical_messagebox.assert_called_once()
+
+        assert_buttons_in_init_state(batch_processing_gui)
+
+
+def test_BatchProcessingGui_clear(qtbot, batch_processing_gui, role_names, filenames):
+    for role_name in role_names:
+        current_data_role_widget = batch_processing_gui._data_role_widgets[role_name]
+        current_data_role_widget.list_widget.addItems(filenames)
+        assert current_data_role_widget.list_widget.count() == len(filenames)
+
+    for role_name in role_names:
+        current_data_role_widget = batch_processing_gui._data_role_widgets[role_name]
+        qtbot.mouseClick(current_data_role_widget.clear_button, Qt.LeftButton, delay=1)
+        assert current_data_role_widget.list_widget.count() == 0
+
+
+@pytest.mark.parametrize(
+    "secondary_files",
+    [tuple(), ("sec1",), ("sec1", "sec2"), ("sec1", "sec2", "sec3"), ("sec1", "sec2", "sec3", "sec4")],
+)
+def test_BatchProcessingGui_nonmatching_raises(qtbot, batch_processing_gui, role_names, filenames, secondary_files):
+    batch_processing_gui._data_role_widgets[role_names[0]].list_widget.addItems(filenames)
+    secondary_data_role = batch_processing_gui._data_role_widgets[role_names[1]]
+    with mock.patch("ilastik.applets.batchProcessing.batchProcessingGui.Request", new=RequestMock):
+        secondary_data_role.list_widget.addItems(secondary_files)
+        if len(secondary_files) > 0 and (len(secondary_files) != len(filenames)):
```

Comment:
```suggestion
        if len(secondary_files) in range(1, len(filenames)):
```

Reply:
hm, actually I want to run this code path whenever there are one or more `secondary_files` except if the length of `secondary_files` _and_ `filenames` are equal. With the range check (which I really like and will definitely add to my active Python vocabulary), but in this case it doesn't help me.

I'll do it like

```python
if len(secondary_files) not in (0, len(filenames)):
...
```

Reply:
In that case, maybe list the failing examples explicitly in parameters?

```python
from contextlib import nullcontext as does_not_raise

from pytest import raises


@pytest.mark.parametrize(
    "secondary_files,expectation",
    [
        ((), does_not_raise()),
        (("sec1",), raises(BatchProcessingDataConstraintException)),
        ...,
    ]
def test_BatchProcessingGui_nonmatching_raises(..., secondary_files, expectation):
    ...
    with expectation:
        qtbot.mouseClick(batch_processing_gui.run_button, Qt.LeftButton, delay=1)
```

http://doc.pytest.org/en/latest/example/parametrize.html#parametrizing-conditional-raising

Reply:
I would give that two thumbs up :)